### DISCUSSION TITLE
Unable to require core NodeJS modules in SSR getInitialProps Code

### DISCUSSION THREAD

**Discussion by ranchodeluxe** [vercel/next.js#31399] [Help]
### What version of Next.js are you using?

9.1.6

### What version of Node.js are you using?

10.18.1

### What browser are you using?

Firefox

### What operating system are you using?

MacOSX BigSure 11.5.2

### How are you deploying your application?

running locally with `next <proj-folder> -p 9213`

### Describe the Bug

I'm using dynamic pages routing and trying to `require` certain experimental and core nodejs libs into my SSR code for `getInitialProps` (see code below). At first I was getting errors for all my required libs, but then I told webpack (in `next.config.js`) to leave these empty for non-server-side bundling. However `async_hooks` still isn't able to be required in `getInitialProps`.

 I keep getting the error: 
`[ error ] ./pages/blahs/details/[blahid].js | Module not found: Can't resolve 'async_hooks' in '/app/blahserver/pages/blahs/details'`

### next.config.js
```javascript
function webpackConfig(config, {isServer}) {
  if (!isServer) {
    config.node = {
      fs: 'empty',
      os: 'empty',
      async_hooks: 'empty',
    };
  }
 // more config below here
}
```

### example [blahid].js page
```javascript
const BlahPage = () => {
  const router = useRouter();
  return (
    <div className='blah-page'>
        ... more stuff here
    </div>
  );
};

BlahPage.getInitialProps = async (ctx) => {
  const fs = require('fs');
  const os = require('os');
  const async_hooks = require('async_hooks');
  // do more stuff here with required libs
}
```

### Expected Behavior

`getInitialProps` should be able to require all core/experimental modules that exist in NodeJS v10.18.1 just fine

### To Reproduce

0. create a dynamic page route
1. write an `async` `getInitialProps` handler that requires some `async_hooks`:

```javascript
Page.getInitialProps = async (ctx) => {
require('async_hooks');
}
```
2. watch it fail
3. profit

**Comment by timneutkens** [comment]
Converted this to discussions "help" section as it is not a bug in Next.js since you've added additional webpack configuration. I'd recommend using `getServerSideProps` which is automatically tree-shaken from the browser bundle making it so that you can import Node.js libraries just fine: https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering

**Reply by ranchodeluxe** [reply]
Thanks for the advice @timneutkens. Unfortunately `getServerSideProps` results in the same error `ModuleNotFoundError: Module not found: Error: Can't resolve 'async_hooks' in '/app/blahserver/pages/blahs/details'`

**Reply by ranchodeluxe** [reply]
@timneutkens: `getServerSideProps` isn't even available in the version of NextJS I'm using (see below). Any other advice?

```bash
https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering
v9.3.0 | getServerSideProps introduced
```

**Reply by oredavids** [reply]
@ranchodeluxe +1. I have a similar issue with using the [cls-rtracer](https://www.npmjs.com/package/cls-rtracer) package with NextJs. The package uses the `AsyncLocalStorage` API from `async_hooks` under the hood.

Did you find a workaround?
